skimage/util/noise.py

```diff
@@ -166,23 +166,18 @@ def random_noise(image, mode='gaussian', seed=None, clip=True, **kwargs):
                            amount=kwargs['amount'], salt_vs_pepper=0.)
 
     elif mode == 's&p':
-        # This mode makes no effort to avoid repeat sampling. Thus, the
-        # exact number of replaced pixels is only approximate.
         out = image.copy()
 
         # Salt mode
-        num_salt = np.ceil(
-            kwargs['amount'] * image.size * kwargs['salt_vs_pepper'])
-        coords = [np.random.randint(0, i - 1, int(num_salt))
-                  for i in image.shape]
-        out[coords] = 1
-
-        # Pepper mode
-        num_pepper = np.ceil(
-            kwargs['amount'] * image.size * (1. - kwargs['salt_vs_pepper']))
-        coords = [np.random.randint(0, i - 1, int(num_pepper))
-                  for i in image.shape]
-        out[coords] = low_clip
+        p = kwargs['amount']
+        q = kwargs['salt_vs_pepper']
+        flipped = np.random.choice([True, False], size=image.shape,
+                                   p=[p, 1 - p])
+        salted = np.random.choice([True, False], size=image.shape,
+                                  p=[q, 1 - q])
```

Comment:
@soupault this time, I know I'm right. =) For small values of p *and* small values of q, you'll get small overlap, and potentially no salt, but that's what a small value of q means. Salting selection is independent of pixel selection, which means the probability of a pixel being salted *given* that it's flipped is exactly the same as the probability of a pixel being salted: q.